### DISCUSSION TITLE
Migration of source code from TFS to Git with all history

### DISCUSSION THREAD

**Discussion by gkrishnat** [community/community#22995] [New to GitHub]
<p>Hi Team,</p>
<p>I wanted to migrate source code from TFS to Git with all history. Kindly provide me the help to accomplish this or if not<br>
all then how old history can be migrated?</p>
<p>Thanks &amp; Regards,<br>
Gopi Thakur</p>

**Comment by seveas** [comment, marked as answer]
<p>Hi <a class="mention" href="/gkrishnat">@gkrishnat</a> </p>
<p>You can use GitHub’s importer at <a href="https://github.com/new/import" rel="nofollow noopener">https://github.com/new/import</a> to import it straight into Github.</p>

**Reply by amolsdalvi** [reply]
How we can import only single branch from Repository?

**Comment by gkrishnat** [comment]
<p>Hi There,</p>
<p>I followed the above link to download git-tfs and was able to install successfully with set path.</p>
<p>But I am getting the below error while migrating using clone command.</p>
<p>Command:- git tfs list-remote-branches <a href="http://tfs.fjt.fawex.com:8080/tfs/WebApp" rel="nofollow noopener">http://tfs.fjt.fawex.com:8080/tfs/WebApp </a></p>
<p>TFS branches that clould be cloned </p>
<p>$/Folder1/Folder2/Project1 [*]</p>
<p>Cloning root branches &lt;marked by [*]&gt; is recommended!</p>
<p>Command:- git tfs clone <a href="http://tfs.fjt.fawex.com:8080/tfs/WebApp" rel="nofollow noopener">http://tfs.fjt.fawex.com:8080/tfs/WebApp</a> “$/Folder1/Folder2/Project1”</p>
<p>Error:- TFS repository can not be root and must start with “$/”.<br>
You may be able to resolve this problem<br>
-Try using $/C:/Program Files/Git/Folder1/Folder2/Project1</p>
<p>​​​​​​​Kindly help me in resolving this.</p>
<p>Thanks &amp; Regards,<br>
Gopi Thakur</p>

**Reply by amolsdalvi** [reply]
Use below command to clone your branch from TFS to Git local folder with history.
tf clone http://tfs-web:8080/tfs/WebApp "$/Repository-Name/Branch-Name" "Local Drive path as like D:\Test" --deep

its works for me.

**Comment by gkrishnat** [comment]
<p>Hi There,</p>
<p>MSYS_NO_PATHCONV=1 git tfs clone <a href="http://tfs.fjt.fawex.com:8080/tfs/WebApp" rel="nofollow noopener">http://tfs.fjt.fawex.com:8080/tfs/WebApp</a> “$/Folder1/Folder2/Project1”  -d<br>
–branches=all --export --export-work-item-mapping=“d:\WorkItems\Mappings\workitems.txt”</p>
<p>The above git tfs command is successfully downloading the solutions in the local git repository. While downloading it is listing all the change sets that are available before migration.</p>
<p>But I faced 2 issues:- </p>
<p>Issue 1:-</p>
<p>In Visual Studio Team Explorer there is an item called Local Git Repositories.It is listing the Local Git Repository Name and I right clicked the repository name where tfs project got migrated.<br>
There were option to “Open Command Prompt”. I clicked that and I got the command prompt with the local repository path listed as a current path. Then I give the command git log --all.<br>
It is first showing the few change sets but after keep on pressing the enter button gives all the change sets that were available before migration. That means all the change sets histories exists in the local path in windows machine.</p>
<p>It is showing the git-tfs-id like below for all the change sets.</p>
<p>git-tfs-id: <a href="http://tfs.fjt.fawex.com:8080/tfs/WebApp%5D%24/Folder1/Folder2/Project1;C1503" rel="nofollow noopener">[http://tfs.fjt.fawex.com:8080/tfs/WebApp]$/Folder1/Folder2/Project1;C1503</a></p>
<p>But in Solution Explorer, opening the .sln file for the same project from the local git repository and then Right click the project name(In the Solution Explorer Window)-&gt; Source Control-&gt; View History. It is showing less change sets.</p>
<p>Issue 2:-</p>
<p>d:\WorkItems\Mappings\workitems.txt is empty.</p>
<p>Thanks &amp; Regards,</p>
<p>Gopi Thakur</p>
